src/main/java/com/google/javascript/clutz/DeclarationGenerator.java

```diff
@@ -2963,9 +2963,7 @@ private String getSignatureForInstanceTTEFn(
         } else {
           return "then < RESULT > (opt_onFulfilled ? : ( (a : "
               + templateVarName
-              + " ) => "
-              + classTemplatizedType
-              + " | RESULT ) | null , "
+              + " ) => PromiseLike < RESULT > | RESULT ) | null , "
```

Comment:
That's a clever solution. In the past we just expected all promise implementations to be structurally compatible, so one can say they expect `then((...): ಠ_ಠ.clutz.goog.Thenable<T>, ...)` and expected that `ಠ_ಠ.clutz.goog.Thenable<T>` is mutually assignable with `PromiseLike<T>`. 

However, we ran into issues when `strictGenericChecks` was on - the two types were not mutually assignable. That's why is is off here and we have it off in the google codebase. I think with your change the types are getting more obviously assignable, so this might no longer be an issue.